File: cc/google/fhir/error_reporter.cc
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "google/fhir/error_reporter.h"

#include <optional>
#include <string>
#include <string_view>

#include "glog/logging.h"
#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"

namespace google::fhir {

const ScopedErrorReporter ScopedErrorReporter::WithScope(
    absl::string_view scope, std::optional<uint> index) const {
  return ScopedErrorReporter(&handler_, scope, index, this);
}

const ScopedErrorReporter ScopedErrorReporter::WithScope(
    const google::protobuf::FieldDescriptor* field,
    std::optional<std::uint8_t> index) const {
  return WithScope(field->json_name(),
                   field->is_repeated() ? index : std::nullopt);
}

std::string ScopedErrorReporter::CurrentFieldPath() const {
  if (prev_scope_ == nullptr) {
    return field_name_;
  }
  return absl::StrCat(prev_scope_->CurrentFieldPath(), ".", field_name_);
}

std::string ScopedErrorReporter::CurrentElementPath() const {
  std::string path_component =
      index_.has_value()
          ? absl::Substitute("$0[$1]", field_name_, index_.value())
          : field_name_;

  if (prev_scope_ == nullptr) {
    return path_component;
  }
  return absl::StrCat(prev_scope_->CurrentElementPath(), ".", path_component);
}

absl::Status ScopedErrorReporter::ReportFhirFatal(
    const absl::Status& status) const {
  return handler_.HandleFhirFatal(status, CurrentElementPath(),
                                  CurrentFieldPath());
}
absl::Status ScopedErrorReporter::ReportFhirFatal(
    const absl::Status& status, absl::string_view field_name,
    std::optional<uint> index) const {
  return WithScope(field_name, index).ReportFhirFatal(status);
}

absl::Status ScopedErrorReporter::ReportFhirError(absl::string_view msg) const {
  return handler_.HandleFhirError(msg, CurrentElementPath(),
                                  CurrentFieldPath());
}
absl::Status ScopedErrorReporter::ReportFhirError(
    absl::string_view msg, absl::string_view field_name,
    std::optional<uint> index) const {
  return WithScope(field_name, index).ReportFhirError(msg);
}

absl::Status ScopedErrorReporter::ReportFhirWarning(
    absl::string_view msg) const {
  return handler_.HandleFhirWarning(msg, CurrentElementPath(),
                                    CurrentFieldPath());
}
absl::Status ScopedErrorReporter::ReportFhirWarning(
    absl::string_view msg, absl::string_view field_name,
    std::optional<uint> index) const {
  return WithScope(field_name, index).ReportFhirWarning(msg);
}

absl::Status ScopedErrorReporter::ReportFhirPathFatal(
    const absl::Status& status, absl::string_view expression) const {
  return handler_.HandleFhirPathFatal(status, expression, CurrentElementPath(),
                                      CurrentFieldPath());
}
absl::Status ScopedErrorReporter::ReportFhirPathFatal(
    const absl::Status& status, absl::string_view expression,
    absl::string_view field_name, std::optional<uint> index) const {
  return WithScope(field_name, index).ReportFhirPathFatal(status, expression);
}

absl::Status ScopedErrorReporter::ReportFhirPathError(
    absl::string_view expression) const {
  return handler_.HandleFhirPathError(expression, CurrentElementPath(),
                                      CurrentFieldPath());
}
absl::Status ScopedErrorReporter::ReportFhirPathError(
    absl::string_view expression, absl::string_view field_name,
    std::optional<uint> index) const {
  return WithScope(field_name, index).ReportFhirPathError(expression);
}

absl::Status ScopedErrorReporter::ReportFhirPathWarning(
    absl::string_view expression) const {
  return handler_.HandleFhirPathWarning(expression, CurrentElementPath(),
                                        CurrentFieldPath());
}
absl::Status ScopedErrorReporter::ReportFhirPathWarning(
    absl::string_view expression, absl::string_view field_name,
    std::optional<uint> index) const {
  return WithScope(field_name, index).ReportFhirPathWarning(expression);
}

FailFastErrorHandler& FailFastErrorHandler::FailOnErrorOrFatal() {
  static FailFastErrorHandler* const kInstance =
      new FailFastErrorHandler(FailFastErrorHandler::FAIL_ON_ERROR_OR_FATAL);
  return *kInstance;
}

FailFastErrorHandler& FailFastErrorHandler::FailOnFatalOnly() {
  static FailFastErrorHandler* const kInstance =
      new FailFastErrorHandler(FailFastErrorHandler::FAIL_ON_FATAL_ONLY);
  return *kInstance;
}

std::optional<uint> IndexOrNullopt(const google::protobuf::FieldDescriptor* field,
                                   uint index) {
  return field->is_repeated() ? std::optional(index) : std::nullopt;
}

}  // namespace google::fhir

File: go/jsonformat/errorreporter/errorreporter.go
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package errorreporter makes all validation errors visible to callers
// so they can report or handle them as appropriate for the surrounding
// system.
package errorreporter

import (
	"fmt"

	"github.com/google/fhir/go/fhirversion"

	c4pb "github.com/google/fhir/go/proto/google/fhir/proto/r4/core/codes_go_proto"
	d4pb "github.com/google/fhir/go/proto/google/fhir/proto/r4/core/datatypes_go_proto"
	r4outcomepb "github.com/google/fhir/go/proto/google/fhir/proto/r4/core/resources/operation_outcome_go_proto"
	c3pb "github.com/google/fhir/go/proto/google/fhir/proto/stu3/codes_go_proto"
	d3pb "github.com/google/fhir/go/proto/google/fhir/proto/stu3/datatypes_go_proto"
	r3pb "github.com/google/fhir/go/proto/google/fhir/proto/stu3/resources_go_proto"
)

// IssueSeverityCode describes the severity of an operation output issue.
type IssueSeverityCode string

// IssueTypeCode describes the type of an operation output issue.
type IssueTypeCode string

// FHIR versions and Issue codes for operation outcome.
const (
	IssueSeverityWarning = IssueSeverityCode("warning")
	IssueSeverityError   = IssueSeverityCode("error")
	ValueIssueTypeCode   = IssueTypeCode("VALUE")
)

var (
	// R3IssueSeverityCodeMap maps IssueSeverityCode to R3IssueSeverityCode
	R3IssueSeverityCodeMap = map[IssueSeverityCode]c3pb.IssueSeverityCode_Value{
		IssueSeverityError:   c3pb.IssueSeverityCode_ERROR,
		IssueSeverityWarning: c3pb.IssueSeverityCode_WARNING,
	}
	// R3OutcomeCodeMap maps IssueTypeCode to R3IssueTypeCode_Value
	R3OutcomeCodeMap = map[IssueTypeCode]c3pb.IssueTypeCode_Value{
		ValueIssueTypeCode: c3pb.IssueTypeCode_VALUE,
	}
	// R4IssueSeverityCodeMap maps IssueSeverityCode to R4IssueSeverityCode
	R4IssueSeverityCodeMap = map[IssueSeverityCode]c4pb.IssueSeverityCode_Value{
		IssueSeverityError:   c4pb.IssueSeverityCode_ERROR,
		IssueSeverityWarning: c4pb.IssueSeverityCode_WARNING,
	}
	// R4OutcomeCodeMap maps IssueTypeCode to R4IssueTypeCode_Value
	R4OutcomeCodeMap = map[IssueTypeCode]c4pb.IssueTypeCode_Value{
		ValueIssueTypeCode: c4pb.IssueTypeCode_VALUE,
	}
)

// An ErrorReporter can be used to handle validation errors in the manner
// of the caller's choosing.
type ErrorReporter interface {
	// ReportValidationError reports validation errors occurred during validation.
	//
	// If the error can be satisfactorily reported it should return nil, instructing
	// the FHIR validation logic to proceed.
	//
	// Conversely, if this returns an error, it means an error was encountered while
	// attempting to handle the original error, indicating a failure mode of the
	// ErrorReporter itself.
	ReportValidationError(elementPath string, err error) error
	// ReportValidationError reports validation warning occurred during validation.
	//
	// If the error can be satisfactorily reported, it should return nil, instructing
	// the FHIR validation logic to proceed.
	//
	// Conversely, if this returns an error, it means an error was encountered while
	// attempting to handle the original error, indicating a failure mode of the
	// ErrorReporter itself.
	ReportValidationWarning(elementPath string, err error) error
}

// MultiVersionOperationOutcome encompasses Operations of multiple FHIR versions.
type MultiVersionOperationOutcome struct {
	Version   fhirversion.Version
	R3Outcome *r3pb.OperationOutcome
	R4Outcome *r4outcomepb.OperationOutcome
}

// OperationErrorReporter is an implementation of ErrorReporter. It makes
// validation errors visible to callers by preserving errors in a
// MultiVersionOperationOutcome.
// TODO: Add examples
type OperationErrorReporter struct {
	Outcome *MultiVersionOperationOutcome
}

// NewOperationErrorReporter returns an OperationErrorRerporter with specified
// fhir version.
func NewOperationErrorReporter(ver fhirversion.Version) *OperationErrorReporter {
	outcome := &MultiVersionOperationOutcome{Version: ver}
	switch ver {
	case fhirversion.STU3:
		outcome.R3Outcome = &r3pb.OperationOutcome{}
	case fhirversion.R4:
		outcome.R4Outcome = &r4outcomepb.OperationOutcome{}
	}
	return &OperationErrorReporter{
		Outcome: outcome,
	}
}

// ReportValidationError reports an issue at "Error" severity, indicating that the operation should
// be considered a failure.
func (oe *OperationErrorReporter) ReportValidationError(elementPath string, err error) error {
	return oe.report(elementPath, err, ValueIssueTypeCode, IssueSeverityError)
}

// ReportValidationWarning reports an issue at "Warning" severity, indicating that the operation
// encountered an issue, but can still be considered successful.
func (oe *OperationErrorReporter) ReportValidationWarning(elementPath string, err error) error {
	return oe.report(elementPath, err, ValueIssueTypeCode, IssueSeverityWarning)
}

func (oe *OperationErrorReporter) report(elementPath string, err error, typeCode IssueTypeCode, severity IssueSeverityCode) error {
	switch oe.Outcome.Version {
	case fhirversion.STU3:
		issues := oe.Outcome.R3Outcome.GetIssue()
		issue := &r3pb.OperationOutcome_Issue{
			Code: &c3pb.IssueTypeCode{
				Value: R3OutcomeCodeMap[typeCode],
			},
			Severity: &c3pb.IssueSeverityCode{
				Value: R3IssueSeverityCodeMap[severity],
			},
			Diagnostics: &d3pb.String{Value: err.Error()},
		}
		if elementPath != "" {
			issue.Expression = append(issue.Expression, &d3pb.String{Value: elementPath})
		}
		oe.Outcome.R3Outcome.Issue = append(issues, issue)
	case fhirversion.R4:
		issues := oe.Outcome.R4Outcome.GetIssue()
		issue := &r4outcomepb.OperationOutcome_Issue{
			Code: &r4outcomepb.OperationOutcome_Issue_CodeType{
				Value: R4OutcomeCodeMap[typeCode],
			},
			Severity: &r4outcomepb.OperationOutcome_Issue_SeverityCode{
				Value: R4IssueSeverityCodeMap[severity],
			},
			Diagnostics: &d4pb.String{Value: err.Error()},
		}
		if elementPath != "" {
			issue.Expression = append(issue.Expression, &d4pb.String{Value: elementPath})
		}
		oe.Outcome.R4Outcome.Issue = append(issues, issue)
	default:
		return fmt.Errorf("unsupported FHIR version %s", oe.Outcome.Version)
	}
	return nil
}

// BasicErrorReporter simply stores all errors during valudation.
//
// This is primarily for legacy use; most users should use an OperationOutcomeErrorReporter
// or their own implementation.
type BasicErrorReporter struct {
	Errors []*error
}

// NewBasicErrorReporter creates a basic error reporter
func NewBasicErrorReporter() *BasicErrorReporter {
	return &BasicErrorReporter{}
}

// ReportValidationError stores the error by appending it to the errors field.
func (be *BasicErrorReporter) ReportValidationError(_ string, err error) error {
	be.Errors = append(be.Errors, &err)
	return nil
}

// ReportValidationWarning does nothing.
func (be *BasicErrorReporter) ReportValidationWarning(elementPath string, err error) error {
	return nil
}
